packages/angular/src/utils/mfe/with-module-federation.ts

```diff
@@ -0,0 +1,238 @@
+  name: string;
+  remotes?: MFERemotes;
+  exposes?: Record<string, string>;
+  shared?: (
+    libraryName: string,
+    library: SharedLibraryConfig
+  ) => SharedLibraryConfig | false;
+}
+
+function recursivelyResolveWorkspaceDependents(
+  projectGraph: ProjectGraph<any>,
+  target: string,
+  collectedDeps: Set<string> = new Set()
+) {
+  if (collectedDeps.has(target)) {
+    return [];
+  }
+  let dependencies = [target];
+
+  const workspaceDependencies = (
+    projectGraph.dependencies[target] ?? []
+  ).filter((dep) => !dep.target.startsWith('npm:'));
+  if (workspaceDependencies.length > 0) {
+    for (const dep of workspaceDependencies) {
+      dependencies = [
+        ...dependencies,
+        ...recursivelyResolveWorkspaceDependents(
+          projectGraph,
+          dep.target,
+          collectedDeps
+        ),
+      ];
+    }
+  }
+  collectedDeps = new Set([...collectedDeps, ...dependencies]);
+  return [...collectedDeps];
+}
+
+function mapWorkspaceLibrariesToTsConfigImport(workspaceLibraries: string[]) {
+  const { projects } = new Workspaces(appRootPath).readWorkspaceConfiguration();
+  const tsConfigPath = process.env.NX_TSCONFIG_PATH ?? getRootTsConfigPath();
+  const tsConfig: ParsedCommandLine = readTsConfig(tsConfigPath);
+
+  const tsconfigPathAliases: Record<string, string[]> = tsConfig.options?.paths;
+
+  if (!tsconfigPathAliases) {
+    return workspaceLibraries;
+  }
+
+  const mappedLibraries = [];
+  for (const lib of workspaceLibraries) {
+    const sourceRoot = projects[lib].sourceRoot;
+    let found = false;
+
+    for (const [key, value] of Object.entries(tsconfigPathAliases)) {
+      if (value.find((p) => p.startsWith(sourceRoot))) {
+        mappedLibraries.push(key);
+        found = true;
+        break;
+      }
+    }
+
+    if (!found) {
+      mappedLibraries.push(lib);
+    }
+  }
+
+  return mappedLibraries;
+}
+
+async function getDependentPackagesForProject(name: string) {
+  let projectGraph: ProjectGraph<any>;
+  try {
+    projectGraph = readCachedProjectGraph();
+  } catch (e) {
+    projectGraph = await createProjectGraphAsync();
+  }
+
+  const deps = projectGraph.dependencies[name].reduce(
+    (dependencies, dependency) => {
+      const workspaceLibraries = new Set(dependencies.workspaceLibraries);
+      const npmPackages = new Set(dependencies.npmPackages);
+
+      if (dependency.target.startsWith('npm:')) {
+        npmPackages.add(dependency.target.replace('npm:', ''));
+      } else {
+        workspaceLibraries.add(dependency.target);
+      }
+
+      return {
+        workspaceLibraries: [...workspaceLibraries],
+        npmPackages: [...npmPackages],
+      };
+    },
+    { workspaceLibraries: [], npmPackages: [] }
+  );
+  deps.workspaceLibraries = deps.workspaceLibraries.reduce(
+    (workspaceLibraryDeps, workspaceLibrary) => [
+      ...workspaceLibraryDeps,
+      ...recursivelyResolveWorkspaceDependents(projectGraph, workspaceLibrary),
+    ],
+    []
+  );
```

Comment:
In addition to the previous comment, to effectively prevent duplicated work, you need to create the `Set` here and keep the instance to pass it to every call to `recursivelyResolveWorkspaceDependents`.

Reply:
Surely the default in the function call does the same job as declaring it here?

Reply:
Oh yeah, across actual iterations of reduce. That’s a very good point.